src/2982R0_std_quantity_as_a_numeric_type.md

```diff
@@ -2284,10 +2287,90 @@ if (is_neq_zero(q1 / q2))
+
+2. **Regretted failures:** these are failures where the user's request is perfectly reasonable and
+   unambiguous: we'd prefer that the user's code work, and do what they expect.
+
+One example of a "non-regretted failure" is a generic speed function, which divides a distance by a time and
+returns the results in the natural derived unit:
+
+```cpp
+QuantityOf<isq::speed> auto average_speed(
+    QuantityOf<isq::distance> auto ds,
+    QuantityOf<isq::time> auto dt)
+{
+    return ds / dt;
+}
+```
+
+In this case, we might be tempted to write `average_speed(ZERO, 10 * si::second)`, but this wouldn't provide
+enough information to determine the units of the return type.  The compilation failure draws our attention to
+this omission.
+
+An example "regretted failure" is the unit-aware `max()` function from [@AU].  This function template takes
+two different quantity types, and returns whichever quantity is larger, expressed in their common unit.  In
+principle, It's perfectly meaningful to pass `Zero` as one of the arguments of `max()`.  (For example,
+a quantity which is known to be non-negative can come from a computation that sometimes return slightly
+negative results, due to floating point error.)  With `max()` it's also perfectly clear (in principle) which
+quantity type we should construct from `Zero`: that of the other argument.  But in _practice_, template
+deduction fails because `Zero` is not itself a quantity.
+
+The remedy is to provide new overloads that handle `Zero`, making the `max()` API feel more seamless and less
+surprising for the end user.  This does have its costs.  For one thing, it creates extra work on the
+implementation side.  It also scales poorly as `Zero` becomes meaningful for more and more parameters in our
+API.  On the other hand, there are hidden benefits as well: if `Zero` becomes a vocabulary type that is useful
+beyond simple quantities, then a single pair of overloads for `max()` will cover every possible type to which
+`Zero` is convertible.
+
+For completeness, we should consider how the special function approach would handle this same case, where one
+of the arguments to `max` is 0.  It fares slightly better on the implementation side, because it requires only
+one new function instead of two: there is only one "slot" to put the nonzero argument.  However, the end user
+experience is worse, because they have to learn another new function name.  Moreover, the best name for that
+function is not immediately clear.  Finally, if there are other types where this "max-with-zero" functionality
+makes sense, we will have to manually add a new overload for every such new type.
+
+Overall, these two approaches --- special functions, and a `Zero` type --- represent two local optima in
+design space.  Each has its strengths and weaknesses; each makes different tradeoffs.  It's currently an open
+question as to which approach would be best suited for a quantity type in the standard library.
```

Comment:
> An example "regretted failure" is the unit-aware `max()` function from [@AU].  This function template takes
two different quantity types, and returns whichever quantity is larger, expressed in their common unit.

> This function template takes two different quantity types

I'd say you'd want `max` constrained with `std::totally_ordered_with`.
In this case, it'd just work with `Zero`.

The parameters don't have to be quantity types.
But you might want at least one to be a quantity type.

> and returns whichever quantity is larger, expressed in their common unit.

This means we want quantities to be `std::common_with<Zero>`.

Reply:
It seems that @JohelEGP could properly constrain `max()` to make it work in such a case, so I will remove that part of the discussion from the paper.